pkg/ddl/notifier/subscribe.go

```diff
@@ -166,61 +166,82 @@ func (n *DDLNotifier) start() {
 	}
 }
 
+// ProcessEventsBatchSize is the number of events to process in a SQL query. It's
+// exposed for testing.
+var ProcessEventsBatchSize = 1024
+
 func (n *DDLNotifier) processEvents(ctx context.Context) error {
-	sysSession, err := n.sysSessionPool.Get()
+	s, err := n.sysSessionPool.Get()
 	if err != nil {
 		return errors.Trace(err)
 	}
-	defer n.sysSessionPool.Put(sysSession)
+	defer n.sysSessionPool.Put(s)
+	sess4List := sess.NewSession(s.(sessionctx.Context))
+	result := n.store.List(ctx, sess4List)
+	defer sess4List.Rollback()
 
-	session := sess.NewSession(sysSession.(sessionctx.Context))
-	changes, err := n.store.List(ctx, session)
+	s2, err := n.sysSessionPool.Get()
 	if err != nil {
 		return errors.Trace(err)
 	}
+	defer n.sysSessionPool.Put(s2)
+	sess4Process := sess.NewSession(s2.(sessionctx.Context))
 
 	// we should ensure deliver order of events to a handler, so if a handler returns
 	// error for previous events it should not receive later events.
 	skipHandlers := make(map[HandlerID]struct{})
-	for _, change := range changes {
-		for handlerID, handler := range n.handlers {
-			if _, ok := skipHandlers[handlerID]; ok {
-				continue
-			}
-			if err2 := n.processEventForHandler(ctx, session, change, handlerID, handler); err2 != nil {
-				skipHandlers[handlerID] = struct{}{}
 
-				if !goerr.Is(err2, ErrNotReadyRetryLater) {
-					logutil.Logger(ctx).Error("Error processing change",
-						zap.Int64("ddlJobID", change.ddlJobID),
-						zap.Int64("multiSchemaChangeSeq", change.multiSchemaChangeSeq),
-						zap.Stringer("handler", handlerID),
-						zap.Error(err2))
+	changes := make([]*SchemaChange, ProcessEventsBatchSize)
+
+	for {
+		count, err2 := result.Read(changes)
```

Comment:
Why not let read return `[]*SchemaChange` directly?
Return `count` and use `changes[:count]` is just a unnecessary detour

Reply:
It's like an `io.Reader` calling convention. Caller can reuse the previous results. Because we are in 1M tables project, I'm afraid that returning 1M `[]*SchemaChange` (it also has internal pointers like `TableInfo`) will add pressure to GC. 

Though it's less common, I think this package / function is the only caller of `ListResult.Read`. Other developers will not modify this package and will not be confused by it.

However I notice that I use `json.Unmarshal` on an non-empty object, not sure the leftover of the object will be overwritten or causes some problems. I'll check it tomorrow.

Reply:
What was your test result?

Reply:
>　pass-in slices so caller can reuse the previous results

Yes, I mean the same thing, just that the code can be write as:

```
func f(changes []*SchemaChange) []*SchemaChange {
     ret := changes[:0]
     ret = append(ret, xxx)
     return ret
}
```

Anyway, this is just some personal taste, so you can choose to accept it or not.

Reply:
> What was your test result?

I didn't test it yet 😂 I hope other colleague can help check it. Theoretically there's no big query now.

> func f(changes []*SchemaChange) []*SchemaChange {

In this new signature, the batch size of one `Read` is not obvious when input is `nil`, or it seems like an `append` where the slice will be grown and batch size is not be limited. Your suggestion is like `EncodeBytes` but here we want to control the batch size as well as reuse buffer.